Outer groups

LtLi0n opened the pull request:
Adding an optional property **Parent** for **GroupAttribute**.

It allows group modules to be within separate files, outside of parent class, while still having a dependency on that parent module.

Tested on my end, works fine, but I may have missed something. 

Please take a closer look.

LtLi0n committed:
Added ParentModule Type to reference other module as parent.

LtLi0n committed:
SearchAsync now returns a tuple, where Item1 is a list of confirmed types without dependencies.
Item2 is a dictionary of key - subtype outside of parent class, value - parent module.

Added ConfirmTypeInfoAsync, because I had to confirm group parent type is valid too.

Added new argument Dictionary<Type, ModuleInfo> moduleDefs to methods:
BuildModule, BuildSubTypes, BuildAsync.

Modules with parent dependencies expect this argument to have value.

LtLi0n committed:
AddModulesAsync logic to include groups with Parent property.

LtLi0n committed:
SearchAsync now returns a tuple of IReadOnlyList<TypeInfo>, IReadOnlyList<TypeInfo> - Dictionary was not very useful to return.

Fixed [DontAutoLoad] to work properly.

Added parent module checking. If it's found as unfit for module creation, it's added to the list, so duplicate checking/errors are avoided.

LtLi0n committed:
Moved most TypeInfo module validation logic from CommandService.

LtLi0n committed:
Exposed ModuleTypeInfo.

BuildModule now throws an exception if outer subtype parent type is not found within moduleDefs.

LtLi0n committed:
Added a check before adding a module manually. Adding outer subtypes now seem to work, but they don't bind with parent modules properly yet.

LtLi0n commented:
Outer subtypes still leave **Parent** module property uninitialized. 
Trying to solve how to bind them properly, so there is a connection.

I should be able to initialize them on autoLoad, but can't figure out binding through manual adding.

I suspect that currently manual adding doesn't work for nested group modules too?

FiniteReality commented:
What the hell is the use case for this?

Does this not work?

```cs
// File1
public partial class X
{
    [Module("derp")]
    public class Y
    {
    }
}

[Module("thing")]
public partial class X
{
    [Module("derpderp")]
    public class Z
    {
    }
}
```

LtLi0n commented:
Seriously?

What if you have a large group module, say for a game within the bot that will have inner 20 submodules or more, so each having their own distinct class and file would be a lot more maintainable. Not to mention little advantages, such as `ctrl` + `left-click` to open parent class when it's not partial.

You tell me, which one looks clearer

this:

```csharp
[Group("game1")]
public partial class Game1Module
{
	[Group("owner")]
	public class Game1OwnerModule
	{
	}
}

public partial class Game1Module
{
	[Group("info")]
	public class Game1InfoModule
	{
	}
}

public partial class Game1Module
{
	[Group("user")]
	public class Game1UserModule
	{
	}
}
```

or this:
```csharp
[Group("game1")]
public class Game1Module
{

}

[Group("owner", Parent = typeof(Game1Module))]
public class Game1OwnerModule
{

}

[Group("user", Parent = typeof(Game1Module))]
public class Game1UserModule
{

}
```

If submodules are expected to contain a lot of code, all I can see are a lot of useless tab indentations judging from the first example.

k-boyle commented:
What happens if you do
```cs
[GroupAttribute("", Parent = typeof(MyOtherModule))]
public class MyModule : ModuleBase 
{

}

[GroupAttribute("", Parent = typeof(MyModule))]
public class MyOtherModule : ModuleBase
{

}
```

Or set the parent to a nested module?

k-boyle commented on src/Discord.Net.Commands/Attributes/GroupAttribute.cs:
```diff
@@ -13,6 +13,11 @@ public class GroupAttribute : Attribute
         public string Prefix { get; }
 
+        /// <summary>
+        ///     Uses this module as parent instead. Allows creation of group modules outside of parent class.
```
as the parent

LtLi0n replied:
Derp, fixed with https://github.com/discord-net/Discord.Net/pull/1399/commits/b693af5a0dd5ef73834b71e3743f462dfee4453b

k-boyle commented on src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs:
```diff
@@ -10,45 +10,59 @@ namespace Discord.Commands
             }
 
-            return result;
+            return Tuple.Create<IReadOnlyList<TypeInfo>, IReadOnlyList<TypeInfo>>(standardModules, queuedModules);
```
value tuple who?

LtLi0n replied:
Unable to use value tuples, for now. More in-depth: #1393

k-boyle commented on src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs:
```diff
@@ -118,6 +134,20 @@ private static void BuildModule(ModuleBuilder builder, TypeInfo typeInfo, Comman
+                        {
+                            if(moduleDefs.ContainsKey(group.ParentModule))
+                            {
+                                foreach (string parentAlias in moduleDefs[group.ParentModule].Aliases)
```
use `TryGetValue` instead of a double lookup

LtLi0n replied:
fixed with https://github.com/discord-net/Discord.Net/pull/1399/commits/eae4839e1ec6d8d82fc61a64696402a0a7f7b1e5

k-boyle commented on src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs:
```diff
@@ -10,45 +10,59 @@ namespace Discord.Commands
 
-        public static async Task<IReadOnlyList<TypeInfo>> SearchAsync(Assembly assembly, CommandService service)
+        ///<summary>Value1 - ready types, Value2 - types dependant on Value1.</summary>
+        public static async Task<Tuple<IReadOnlyList<TypeInfo>, IReadOnlyList<TypeInfo>>> SearchAsync(Assembly assembly, CommandService service)
```
value tuple

LtLi0n replied:
Unable to use value tuples, for now. More in-depth: #1393

k-boyle commented:
> Seriously?
> 
> What if you have a large group module, say for a game within the bot that will have inner 20 submodules or more, so each having their own distinct class and file would be a lot more maintainable. Not to mention little advantages, such as `ctrl` + `left-click` to open parent class when it's not partial.
> 
> You tell me, which one looks clearer
> 
> this:
> 
> ```cs
> [Group("game1")]
> public partial class Game1Module
> {
> 	[Group("owner")]
> 	public class Game1OwnerModule
> 	{
> 	}
> }
> 
> public partial class Game1Module
> {
> 	[Group("info")]
> 	public class Game1InfoModule
> 	{
> 	}
> }
> 
> public partial class Game1Module
> {
> 	[Group("user")]
> 	public class Game1UserModule
> 	{
> 	}
> }
> ```
> 
> or this:
> 
> ```cs
> [Group("game1")]
> public class Game1Module
> {
> 
> }
> 
> [Group("owner", Parent = typeof(Game1Module))]
> public class Game1OwnerModule
> {
> 
> }
> 
> [Group("info", Parent = typeof(Game1Module))]
> public class Game1InfoModule
> {
> 
> }
> 
> [Group("user", Parent = typeof(Game1Module))]
> public class Game1UserModule
> {
> 
> }
> ```
> 
> If submodules are expected to contain a lot of code, all I can see are a lot of useless tab indentations judging from the first example.

Adding an overload to `ModuleBuilder#AddModule` to take a type and then override `OnModuleBuilding` and adding the modules that way would prob be cleaner ngl

LtLi0n commented:
Thank you for your review. 
I appreciate your effort and you stated many valid arguments with some test cases to keep in mind.

> What happens if you do
> 
> ```cs
> [GroupAttribute("", Parent = typeof(MyOtherModule))]
> public class MyModule : ModuleBase 
> {
> 
> }
> 
> [GroupAttribute("", Parent = typeof(MyModule))]
> public class MyOtherModule : ModuleBase
> {
> 
> }
> ```
> 
> Or set the parent to a nested module?

The above example throws an exception, because parent was not found within a dictionary of regular modules. 
Nested submodules don't work for that exact same reason, good observation.
I really need to join modules before they get built, then cycle through them until it encounters a circular dependency or it's finished. Would solve parent binding too, so **Module.Parent** and **Module.SubModules** properties get initialized.

Will try to resolve:
- [ ] circular dependencies
- [ ] nested outer subtype parent support 
- [ ] proper parent links

> > Seriously?
> > What if you have a large group module, say for a game within the bot that will have inner 20 submodules or more, so each having their own distinct class and file would be a lot more maintainable. Not to mention little advantages, such as `ctrl` + `left-click` to open parent class when it's not partial.
> > You tell me, which one looks clearer
> > this:
> > ```cs
> > [Group("game1")]
> > public partial class Game1Module
> > {
> > 	[Group("owner")]
> > 	public class Game1OwnerModule
> > 	{
> > 	}
> > }
> > 
> > public partial class Game1Module
> > {
> > 	[Group("info")]
> > 	public class Game1InfoModule
> > 	{
> > 	}
> > }
> > 
> > public partial class Game1Module
> > {
> > 	[Group("user")]
> > 	public class Game1UserModule
> > 	{
> > 	}
> > }
> > ```
> > 
> > 
> > or this:
> > ```cs
> > [Group("game1")]
> > public class Game1Module
> > {
> > 
> > }
> > 
> > [Group("owner", Parent = typeof(Game1Module))]
> > public class Game1OwnerModule
> > {
> > 
> > }
> > 
> > [Group("info", Parent = typeof(Game1Module))]
> > public class Game1InfoModule
> > {
> > 
> > }
> > 
> > [Group("user", Parent = typeof(Game1Module))]
> > public class Game1UserModule
> > {
> > 
> > }
> > ```
> > 
> > 
> > If submodules are expected to contain a lot of code, all I can see are a lot of useless tab indentations judging from the first example.
> 
> Adding an overload to `ModuleBuilder#AddModule` to take a type and then override `OnModuleBuilding` and adding the modules that way would prob be cleaner ngl

How come? The user would have to manually add each outer subtype then.
That would create extra work for a user when they decide to add, rename or remove outer subtype classes.

k-boyle commented:
Huh? It's as much work as specify a group on each module and its parent.
Plus afaik the implementation within the library should already but there,
just need to expose a method

On Thu, 17 Oct 2019, 22:24 Saulius Šaltenis, <notifications@github.com>
wrote:

> Thank you for your review.
> I appreciate your effort and you stated many valid arguments with some
> test cases to keep in mind.
>
> What happens if you do
>
> [GroupAttribute("", Parent = typeof(MyOtherModule))]public class MyModule : ModuleBase
> {
>
> }
>
> [GroupAttribute("", Parent = typeof(MyModule))]public class MyOtherModule : ModuleBase
> {
>
> }
>
> Or set the parent to a nested module?
>
> The above example throws an exception, because parent was not found within
> a dictionary of regular modules.
> Nested submodules don't work for that exact same reason, good observation.
> I really need to join modules before they get built, then cycle through
> them until it encounters a circular dependency or it's finished. Would
> solve parent binding too, so *Module.Parent* and *Module.SubModules*
> properties get initialized.
>
> Will try to resolve:
>
>    - circular dependencies
>    - nested outer subtype parent support
>    - proper parent links
>
> Seriously?
> What if you have a large group module, say for a game within the bot that
> will have inner 20 submodules or more, so each having their own distinct
> class and file would be a lot more maintainable. Not to mention little
> advantages, such as ctrl + left-click to open parent class when it's not
> partial.
> You tell me, which one looks clearer
> this:
>
> [Group("game1")]public partial class Game1Module
> {
> 	[Group("owner")]
> 	public class Game1OwnerModule
> 	{
> 	}
> }
> public partial class Game1Module
> {
> 	[Group("info")]
> 	public class Game1InfoModule
> 	{
> 	}
> }
> public partial class Game1Module
> {
> 	[Group("user")]
> 	public class Game1UserModule
> 	{
> 	}
> }
>
> or this:
>
> [Group("game1")]public class Game1Module
> {
>
> }
>
> [Group("owner", Parent = typeof(Game1Module))]public class Game1OwnerModule
> {
>
> }
>
> [Group("info", Parent = typeof(Game1Module))]public class Game1InfoModule
> {
>
> }
>
> [Group("user", Parent = typeof(Game1Module))]public class Game1UserModule
> {
>
> }
>
> If submodules are expected to contain a lot of code, all I can see are a
> lot of useless tab indentations judging from the first example.
>
> Adding an overload to ModuleBuilder#AddModule to take a type and then
> override OnModuleBuilding and adding the modules that way would prob be
> cleaner ngl
>
> How come? The user would have to manually add each outer subtype then.
> That would create extra work for a user when they decide to add, rename or
> remove outer subtype classes.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/discord-net/Discord.Net/pull/1399?email_source=notifications&email_token=AECSRDOHW4GYFLSEB4EWPDDQPDJZ5A5CNFSM4JBSA252YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBRSIXA#issuecomment-543368284>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AECSRDICZ4UCLEUDGHIATVDQPDJZ5ANCNFSM4JBSA25Q>
> .
>

LtLi0n committed:
Merge pull request #1 from discord-net/dev

merging recent .NET Core 3.0 support

LtLi0n committed:
Fixing double dictionary lookup.

LtLi0n committed:
Fixed a grammar mistake.
Added a paragraph explaining the use case for outer group submodules.

LtLi0n committed:
Merge branch 'outer-groups' of https://github.com/LtLi0n/Discord.Net into outer-groups

LtLi0n committed:
Parent modules are now correctly bounded with outer subtypes.

Tuples are changed to ValueTuples due to a recent legacy support drop.

ModuleClassBuilder.SearchAsync now returns an IReadOnlyDictionary<TypeInfo, List<TypeInfo>> where List<TypeInfo> contains types that key is the owner of.
(In other words, indexing with typeinfo returns outer subtypes of that parent).

LtLi0n commented:
```csharp
[Group("panel")]
public class PanelModule : ModuleBase<SocketCommandContext>
{
    [Command("test")]
    public async Task Test() => await ReplyAsync("panel here");
}

[Group("settings", ParentModule = typeof(PanelModule))]
public class PanelSettingsModule : ModuleBase<SocketCommandContext>
{
    [Command("test")]
    public async Task Test() => await ReplyAsync("settings here");
}

[Group("guild", ParentModule = typeof(PanelSettingsModule))]
public class PanelSettingsGuildModule : ModuleBase<SocketCommandContext>
{
    [Command("test")]
    public async Task Test() => await ReplyAsync("settings:guild here");
}

[Group("info", ParentModule = typeof(PanelModule))]
public class PanelInfoModule : ModuleBase<SocketCommandContext>
{
    [Command("test")]
    public async Task Test() => await ReplyAsync("info here");

    [Group("nested")]
    public class PanelInfoNestedModule : ModuleBase<SocketCommandContext>
    {
        [Command("test")]
        public async Task Test() => await ReplyAsync("nested info here");
    }
}
```

### output:

![image](https://user-images.githubusercontent.com/31516441/67896232-9cf34200-fb64-11e9-993d-f5b1557b59d7.png)

LtLi0n committed:
Modules that contain inner nested and outer subtypes are now loaded properly.

LtLi0n reviewed (commented):
Fixes a priority bug for outer subtypes, where inner subtypes would be ignored if outer subtypes were also present.

```csharp
public class TestModule : ModuleBase<SocketCommandContext>
{
    [Group("test")]
    public class TestWithin : ModuleBase<SocketCommandContext>
    {
        [Group("inner")]
        public class InnerModule : ModuleBase<SocketCommandContext>
        {
            [Command]
            public async Task Test() => await ReplyAsync("inner hello");
        }
    }
}

[Group("outer", ParentModule = typeof(TestModule.TestWithin))]
public class OuterModule : ModuleBase<SocketCommandContext>
{
    [Command]
    public async Task Test() => await ReplyAsync("outer hello");
}
```

### output:
![image](https://user-images.githubusercontent.com/31516441/67944348-edaa7f80-fbe4-11e9-8082-5a4727d753a3.png)

quinchs commented:
@Cenngo Could you take a look at this?

Cenngo commented:
> @Cenngo Could you take a look at this?

Honestly i dont see why this is should preferred over partial classes. with partial classes you can have shared properties, methods etc. If the main objective is to be able to seperate modules into multiple files, partial classes should cover that.
